Update EIP-2565 to clarify specification (#2892)
ethereum/EIPs: EIPS/eip-2565.md

title: Repricing of the EIP-198 ModExp precompile
author: Kelly Olson (@ineffectualproperty), Sean Gulley (@sean-sn), Simon Peffers (@simonatsn), Justin Drake (@justindrake), Dankrad Feist (@dankrad)
discussions-to: https://ethereum-magicians.org/t/big-integer-modular-exponentiation-eip-198-gas-cost/4150
status: Accepted
type: Standards Track
category: Core
created: 2020-03-20
requires: 198
---

## Simple Summary

> The [EIP-198](eip-198.md) ‘big integer modular exponentiation’, or `ModExp`, precompile is currently overpriced. Re-pricing this precompile will enable more cost efficient verification of RSA signatures, verifiable delay functions (VDFs), primality checks, and more.

Reviewer: ```suggestion
The [EIP-198](./eip-198.md) ‘big integer modular exponentiation’, or `ModExp`, precompile is currently overpriced. Re-pricing this precompile will enable more cost efficient verification of RSA signatures, verifiable delay functions (VDFs), primality checks, and more.
```

author: Kelly Olson (@ineffectualproperty), Sean Gulley (@sean-sn), Simon Peffers (@simonatsn), Justin Drake (@justindrake), Dankrad Feist (@dankrad)
discussions-to: https://ethereum-magicians.org/t/big-integer-modular-exponentiation-eip-198-gas-cost/4150
status: Last Call
review-period-end: 2020-10-09
type: Standards Track
category: Core
created: 2020-03-20
requires: 198
---

## Simple Summary

> This EIP reprices the [EIP-198](./eip-198.md) ‘big integer modular exponentiation’, or `ModExp`, precompile.

Reviewer: ```suggestion
Defines the gas cost of the `ModExp` (`0x00..05`) precompile.
```
As mentioned previously, we want this entire EIP to stand almost entirely on its own.  It still needs to have a dependency on 198 for the definition of what the ModExp precompile does, but it shouldn't depend on 198 for any of the gas cost stuff being proposed here.

## Abstract
To accurately reflect the real world operational cost of the `ModExp` precompile, this EIP specifies an algorithm for calculating the gas cost. This algorithm approximates the multiplication complexity cost and multiplies that by an approximation of the iterations required to execute the exponentiation.

## Motivation
Modular exponentiation is a foundational arithmetic operation for many cryptographic functions including signatures, VDFs, SNARKs, accumulators, and more. Unfortunately, the ModExp precompile is currently over-priced, making these operations inefficient and expensive. By reducing the cost of this precompile, these cryptographic functions become more practical, enabling improved security, stronger randomness (VDFs), and more.

## Specification
As of `FORK_BLOCK_NUMBER`, the gas cost of calling the precompile at address `0x0000000000000000000000000000000000000005` will be calculated as follows:
```
def calculate_multiplication_complexity(base_length, modulus_length):
    max_length = max(base_length, modulus_length)
    words = math.ceil(max_length / 8)
    return words**2

> def calculate_iteration_count(exponent_length, exponent): # same as ADUSTED_EXPONENT_LENGTH in EIP-198

Reviewer: ```suggestion
def calculate_iteration_count(exponent_length, exponent):
```
Since this EIP stands on its own, no need to reference 198 here as it could lead to confusion (it confused me at least).

Author: @MicahZoltu I added this to specifically to help implementers who are migrating from EIP-198 to EIP-2565, which would be the case at least for Ethereum mainnet. I'm happy to delete if it is confusing. @holiman previously I had structured this EIP as changes that should be made on top of EIP-198. We've since changed it to be more of a 'stand alone' EIP that could be adopted in lieu of EIP-198's gas pricing. It would be great to hear your perspective on if these sort of notes are useful for implementers or only add confusion.

```
def mult_complexity(x):
    ceiling(x/8)^2
```
where is `x` is `max(length_of_MODULUS, length_of_BASE)`. `x` is divided by 8 to account for the number of limbs in multiprecision arithmetic.

### 2. Change value of `GQUADDIVISOR`
`GQUADDIVISOR` is set to `20` per EIP-198. We recommend changing the value of this parameter to `3` to account for the changes in the recommended 'computational complexity' formula above.

### 3. Set a minimum price for calling the precompile

> We recommend wrapping the entire function with a minimum gas price of 200 to ensure that a minimum amount of gas is used when the precompile is called e.g. `max(200,floor(mult_complexity(max(length_of_MODULUS, length_of_BASE)) * max(ADJUSTED_EXPONENT_LENGTH, 1) / GQUADDIVISOR))`

Reviewer 2: ```suggestion
The entire function is wrapped, to ensure a minimum gas price of `200`, i.e:
 `
max(200,floor(mult_complexity(max(length_of_MODULUS, length_of_BASE)) * max(ADJUSTED_EXPONENT_LENGTH, 1) / GQUADDIVISOR))
`
```

The biggest security consideration for this EIP is creating a potential DoS vector by making ModExp operations too inexpensive relative to their computation time.

## References

> [EIP-198](eip-198.md) 

Reviewer: ```suggestion
[EIP-198](./eip-198.md) 
```
